sound/soc/sof/sof-audio.c

```diff
@@ -310,7 +310,7 @@ sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widg
 	struct snd_soc_dapm_path *p;
 
 	/* return if the widget is in use or if it is already unprepared */
-	if (!swidget->prepared || swidget->use_count > 1)
+	if (!swidget->prepared || swidget->use_count > 0)
```

Comment:
sorry this makes no sense to me @bardliao @ranj063 

You want to return here if the widget has two users. If there is a single user you still want to unprepare.
This was added in c02c3b653c7342fe0e21c7cd918a7905a92486d6 by @RanderWang for mixing usages.

In addition, in the sof_prepare_widgets_path we only use swidget->prepared, that's unbalanced and makes no sense either.  If we use the use count on unprepare we should also use it on prepare, no?

Reply:
oh it was @RanderWang and not me. I was wondering why I added this and couldnt remember the reason. I should have checked git blame. But let's think about when unprepare is called. Its called after all the widgets in the DAPM widget list are freed. So a widget should be unprepared only if there are no users, right? So use_count > 0 makes sense.

Reply:
@bardliao my understanding is that we follow the ALSA logic of open/prepare/unprepare/free. I don't see how we could unprepare AFTER the free.

Reply:
but this unprepare() isnt related to the ALSA ioctls at all @plbossart. We do a prepare followed by setup() for all widgets during hw_params and then a free followed by unprepare during hw_free. Have a look at https://github.com/thesofproject/linux/blob/929cd48102671d515a484c6da284d14bd793dff9/sound/soc/sof/sof-audio.c#L636

Reply:
A free followed by unprepare? how is this even possible. Wow, we have all the naming backwards I am afraid.

Reply:
yes, I think we do have the namig inverted. The problem was setup() came first and for lack of a better name, I went with prepare(). But ideally, the prepare should be called setup and the setup() should be called init(). The we'd have setup() followed by init() and then free() followed by reset() maybe?

Reply:
we need to rename all this:
sof_widget_list_setup -> sof_widget_list_prepare_setup
sof_widget_list_free -> sof_widget_list_free_unprepare

Reply:
yes but going back to the original question. I think the logic is correct with use_count > 0 right?

Reply:
Yes indeed, the unprepare is when there are no users left with your definition.

Reply:
In original code, swidget->use_count = 2 when the widget was unprepared. Now swidget->use_count = 1 in this function with many change in kernel.